phoenix-core/src/main/java/org/apache/phoenix/parse/LiteralParseNode.java

```diff
@@ -80,4 +80,35 @@ public boolean isStateless() {
     public String toString() {
         return type == PVarchar.INSTANCE ? ("'" + value.toString() + "'") : value == null ? "null" : value.toString();
     }
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((type == null) ? 0 : type.hashCode());
+		result = prime * result + ((value == null) ? 0 : value.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		LiteralParseNode other = (LiteralParseNode) obj;
+		if (type == null) {
```

Comment:
Instead of comparing type and value for equals, do this:
    return type.isComparableTo(other.type) && type.compareTo(value, other.value, other.type) == 0;

This will ensure that a CHAR type of 'AAA' will equal a VARCHAR type of 'AAA' and a DECIMAL type with a value of 5.0 will equal an INTEGER type with a value of 5.